jsonrpsee/src/macros.rs

```diff
@@ -0,0 +1,116 @@
+macro_rules! cfg_client {
+    ($($item:item)*) => {
+        $(
+            #[cfg(any(feature = "jsonrpsee-http-client", feature = "jsonrpsee-ws-client", feature = "client", feature = "async-client"))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_http_client {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-http-client")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-http-client")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_ws_client {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-ws-client")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-ws-client")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_wasm_client {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-wasm-client")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-wasm-client")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_async_client {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "async-client")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "async-client")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_client_transport {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-client-transport")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-client-transport")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_server {
+    ($($item:item)*) => {
+        $(
+            #[cfg(any(feature = "jsonrpsee-http-server", feature = "jsonrpsee-ws-server"))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_http_server {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-http-server")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-http-server")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_ws_server {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-ws-server")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-ws-server")))]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_proc_macros {
+    ($($item:item)*) => {
+        $(
+            #[cfg(feature = "jsonrpsee-proc-macros")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "jsonrpsee-proc-macros")))]
+            $item
+        )*
+    }
+}
```

Comment:
Would it make sense to write a macro like:

```rust
macro_rules! cfg_feature {
    ($feature:literal, $($item:item)*) => {
        $(
            #[cfg(feature = $feature)]
            #[cfg_attr(docsrs, doc(cfg(feature = $feature)))]
            $item
        )*
    }
}
```

And then all of the other macros across these crates can just call that internally like:

```rust
macro_rules! cfg_proc_macros {
    ($($item:item)*) => cfg_feature!("jsonrpc-proc-macros", $($item:item)*)
}
```

There seem like enough copies to justify it :)

Reply:
yes, I agree